File: 2024/Day 05/Python/main.py
"""Advent of Code - 2024 - Day 5"""
from functools import cmp_to_key

def get_cmp(page_rules: list[list[int]]):
    """returns a comparison function based on the page rules"""
    def cmp(a: int, b: int):
        for (l, r) in page_rules:
            if a == l and b == r:
                return 1
            if b == l and a == r:
                return -1
        return 0
    return cmp

with open(0, encoding="utf-8") as f:
    rules, updates = f.read().split("\n\n", 1)
rules = [list(map(int, rule.split("|"))) for rule in rules.splitlines()]
updates = [list(map(int, update.split(","))) for update in updates.splitlines()]

sum1: int = 0
sum2: int = 0
for u in updates:
    mid = len(u)//2
    if not any(l in u and r in u and u.index(l) >= u.index(r) for (l,r) in rules):
        sum1 += u[mid]
    else:
        u.sort(key=cmp_to_key(get_cmp(rules)))
        sum2 += u[mid]
print(sum1)
print(sum2)

File: 2024/Day 05/Rust/src/main.rs
use std::io;
use std::cmp::Ordering;

fn get_cmp(rules: Vec<(i32, i32)>) -> impl Fn(&i32, &i32) -> Ordering {
    move |a:&i32, b:&i32|
    if rules.contains(&(*a, *b)) { Ordering::Less }
    else if rules.contains(&(*b, *a)) { Ordering::Greater }
    else {Ordering::Equal}
}

fn main() {
    let binding = io::stdin()
        .lines()
        .collect::<Result<Vec<String>, io::Error>>()
        .unwrap()
        .join("\n");
    let mut input = binding.split("\n\n");
    let rules: Vec<(i32, i32)> = input
        .next()
        .unwrap()
        .split("\n")
        .map(|rule| rule.split("|").map(|r| r.parse::<i32>().unwrap()))
        .map(|mut r| (r.next().unwrap(), r.next().unwrap()))
        .collect();
    let updates: Vec<Vec<i32>> = input
        .next()
        .unwrap()
        .split("\n")
        .map(|rule| rule.split(",").map(|r| r.parse::<i32>().unwrap()).collect())
        .collect();

    let cmp = get_cmp(rules.clone());

    let mut sum1 = 0;
    let mut sum2 = 0;
    for mut update in updates {
        let mid = (update.len() - 1) / 2;
        if rules
            .iter()
            .map(|(l, r)| {
                !(update.contains(l) && update.contains(r) && !(update.iter().position(|x| x == l) < update.iter().position(|x| x == r)))
            })
            .all(|x| x)
        {
            sum1 += update[mid];
        }
        else
        {
            update.sort_unstable_by(&cmp);
            sum2 += update[mid];
        }
    }
    println!("{}", sum1);
    println!("{}", sum2);
}
